R/aggregate.R

```diff
@@ -0,0 +1,166 @@
+##' This function calculates the robust summarisation for each feature
+##' (protein). Note that the function assumes that the intensities in
+##' input `e` are already log-transformed.
+##'
+##' @title Return the Robust Expression Summary of a matrix
+##' @param x A feature by sample `matrix` containing quantitative
+##'     data with mandatory `colnames` and `rownames`.
+##' @param ... Additional arguments passed to [MASS::rlm()].
+##' @return `numeric()` vector of length `ncol(x)` with robust
+##'     summarised values.
+##' 
+##' @author Adriaan Sticker, Sebastian Gibb and Laurent Gatto
+##' 
+##' @export
+##' 
+##' @importFrom MASS rlm
+##' 
+##' @examples
+##' x <- matrix(rnorm(30), nrow = 3)
+##' colnames(x) <- letters[1:10]
+##' rownames(x) <- LETTERS[1:3]
+##' robustSummary(x)
+robustSummary <- function(x, ...) {
+    stopifnot(!is.null(colnames(x)))
+    stopifnot(!is.null(rownames(x)))
+    ## If there is only one 1 peptide for all samples return
+    ## expression of that peptide
+    if (nrow(x) == 1L) return(x)
+
+    ## remove missing values
+    p <- !is.na(x)
+    expression <- x[p] ## expression becomes a vector
+    sample <- rep(colnames(x), each = nrow(x))[p]
+    feature <- rep(rownames(x), times = ncol(x))[p]
+
+    ## model.matrix breaks on factors with 1 level so make vector of
+    ## ones (intercept).
+    if (length(unique(sample)) == 1L) sample <- rep(1, length(sample))
+
+    ## Sum contrast on peptide level so sample effect will be mean
+    ## over all peptides instead of reference level.
+    X <- stats::model.matrix(~ -1 + sample + feature,
+                             contrasts.arg = list(feature = 'contr.sum'))
+    ## MASS::rlm breaks on singulare values.
+    ## - Check with base lm if singular values are present.
+    ## - If so, these coefficients will be zero, remove this collumn
+    ##   from model matrix
+    ## - Rinse and repeat on reduced modelmatrx till no singular
+    ##   values are present
+    repeat {
+        fit <- stats::.lm.fit(X, expression)
+        id <- fit$coefficients != 0
+        X <- X[ , id, drop = FALSE]
+        if (!any(!id)) break
+    }
+    ## Last step is always rlm: calculate estimated effects effects as
+    ## summarised values
+    fit <- MASS::rlm(X, expression, ...)
+
+    sampleid <- seq_along(unique(sample))
+    ## This will be needed for NUSE-type of quality control, but will
```

Comment:
Couldn't this comment be removed?

Reply:
This would be an interesting QC for aggregation. I agree it would be better in an issue, but I feel that at this stage, it would be out of scope. Can we just leave it for now.